src/libraries/Microsoft.PowerFx.Core/Public/Values/InMemoryRecordValue.cs

```diff
@@ -41,5 +43,29 @@ protected override bool TryGetField(FormulaType fieldType, string fieldName, out
         {
             return _fields.TryGetValue(fieldName, out result);
         }
+
+        public override FormulaValue UpdateFields(IEnumerable<KeyValuePair<string, FormulaValue>> record)
+        {
+            if (_fieldsDict == null || _fieldsDict.IsReadOnly)
```

Comment:
> || _fieldsDict.IsReadOnly [](http://example.com/codeflow?start=35&length=26)

to consider:
we could check this once in the ctor. 

if (_fieldsDict.IsReadOnly) {
 _fieldsDict = null;
}

And then elsewhere we're just doing
if ( _fields == null)